osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java

```diff
@@ -71,29 +71,79 @@ public InstallationService() {
                 deviceMessageMetadata.getMessageType(), deviceMessageMetadata.getMessagePriority(),
                 deviceMessageMetadata.getScheduleTime());
     }
 
+    public void updateSubscriptionInformation(final DeviceMessageMetadata deviceMessageMetadata,
+            final SetSubscriptionInformationRequestData requestData) throws FunctionalException {
+
+        final String organisationId = deviceMessageMetadata.getOrganisationIdentification();
+        final String deviceIdentification = deviceMessageMetadata.getDeviceIdentification();
+
+        final SmartMeter updatedSmartMeter = this.smartMeterService.updateSubscriptionInformation(
+                deviceMessageMetadata.getDeviceIdentification(), requestData.getIpAddress(), requestData.getBtsId(),
+                requestData.getCellId());
+
+        log.debug("updateSubscriptionInformation for organisationIdentification: {} for deviceIdentification: {}. "
+                + "New ipAdress = {} ", organisationId, deviceIdentification, updatedSmartMeter.getIpAddress());
+
+        final SetSubscriptionInformationResponseData responseData = this.getSetSubscriptionInformationResponseData(
+                requestData, updatedSmartMeter);
+
+        final ResponseMessage responseMessage = ResponseMessage.newResponseMessageBuilder()
+                .withCorrelationUid(deviceMessageMetadata.getCorrelationUid())
+                .withOrganisationIdentification(deviceMessageMetadata.getOrganisationIdentification())
+                .withDeviceIdentification(deviceMessageMetadata.getDeviceIdentification())
+                .withResult(ResponseMessageResultType.OK)
+                .withDataObject(responseData)
+                .withMessagePriority(deviceMessageMetadata.getMessagePriority())
+                .build();
+
+        this.webServiceResponseMessageSender.send(responseMessage, deviceMessageMetadata.getMessageType());
+    }
+
+    private SetSubscriptionInformationResponseData getSetSubscriptionInformationResponseData(
+            final SetSubscriptionInformationRequestData requestData, final SmartMeter updatedSmartMeter) {
+        final SetSubscriptionInformationResponseData responseData = new SetSubscriptionInformationResponseData();
+        responseData.setMeId(requestData.getMeId());
+        responseData.setEsn(requestData.getEsn());
+        responseData.setUimId(requestData.getUimId());
+        responseData.setEqId(requestData.getEqId());
+        responseData.setIpAddress(updatedSmartMeter.getIpAddress());
+        responseData.setMdn(requestData.getMdn());
+        responseData.setBtsId(updatedSmartMeter.getBtsId());
+        responseData.setCellId(updatedSmartMeter.getCellId());
+        responseData.setStatus(requestData.getStatus());
+        responseData.setCustCode(requestData.getCustCode());
+        responseData.setSupplierReferenceId(requestData.getSupplierReferenceId());
```

Comment:
Yes, the scenario here is a 2-phase action
- call from SMHE recipe to get the CDMA info from zSmart.
- on success, the data should be pushed to GXF. 
- on success, the entire set of data retrieved from zSmart should be shown in SMHE after the Device was successfully updated.

This means I have to round trip all of the data because of the asynchronous character of SMHE - GXF communication.

There is no way to save the data to be shown in SMHE for only the first step and then continue with the next step. The data is saved via ResponseProcessors that act upon a Response message from GXF.

Reply:
Gerbert agreed to push and return only the relevant values.

Reply:
Based on my comment, and your final reaction about only pushing and returning relevant values, I would expect most of these lines to have been removed.
Apparently, at least `SetCommunicationNetworkInformationRequestData` and `SetCommunicationNetworkInformationResponseData` still have fields that I no longer expect to see (all those echoed back from `requestData` instead of from the `updatedSmartMeter`).

Reply:
you are right, I missed those.